infrastructure/filefilter/filter.go

```diff
@@ -55,38 +55,40 @@ func (f *fileFilter) findNonIgnoredFiles() <-chan string {
 			}
 
 			if dirEntry.IsDir() {
-				// The following optimization has been removed because it doesn't take into account negation rules:
-				// For example:
-				// .ignoreme
-				// !.ignoreme/keepme.txt
-				//
-				// This implementation would skip the entire folder and not return keepme.txt:
-				//
-				// ```
-				//parentFolderPath := filepath.Dir(path)
-				//if path != rootFolder && ignoreCheckerPerFolder[parentFolderPath].MatchesPath(path) {
-				//	return filepath.SkipDir
-				//}
-				// ```
 				globs := f.collectGlobs(path)
 				f.globsPerFolder[path] = globs
-				f.ignoreCheckerPerFolder[path] = ignore.CompileIgnoreLines(globs...)
 				return nil
 			} else {
 				folderPath := filepath.Dir(path)
-				checker := f.ignoreCheckerPerFolder[folderPath]
-				wg.Add(1)
-				go func() {
-					defer wg.Done()
-					if !checker.MatchesPath(path) {
-						resultsCh <- path
-					}
-				}()
+				filesPerFolder[folderPath] = append(filesPerFolder[folderPath], path)
 			}
 
 			return nil
 		})
 
+		sem := make(chan any, 20)
```

Comment:
Replace hard coded value (probably with CPU count)

Reply:
Never assume to be allowed to take all the CPUs... better, just restrict it to 2 or 3 to be easy on users resources.